pymc/distributions/dist_math.py

```diff
@@ -115,6 +117,47 @@ def std_cdf(x):
     return 0.5 + 0.5 * pt.erf(x / pt.sqrt(2.0))
 
 
+def truncated_logcdf_from_base_rv(base_rv_op, value, lower, upper, *rv_inputs):
+    logcdf = _logcdf(base_rv_op, value, *rv_inputs)
+
+    # For left truncated discrete RVs, we don't want to include the lower bound in the
+    # normalization term
+    lower_value = lower - 1 if base_rv_op.dtype.startswith("int") else lower
+    lower_logcdf = _logcdf(base_rv_op, lower_value, *rv_inputs)
+    upper_logcdf = _logcdf(base_rv_op, upper, *rv_inputs)
+
+    is_lower_bounded = not (isinstance(lower, TensorConstant) and np.all(np.isneginf(lower.value)))
+    is_upper_bounded = not (isinstance(upper, TensorConstant) and np.all(np.isinf(upper.value)))
+
+    lognorm = 0
+    if is_lower_bounded and is_upper_bounded:
+        lognorm = logdiffexp(upper_logcdf, lower_logcdf)
+    elif is_lower_bounded:
+        lognorm = pt.log1mexp(lower_logcdf)
+    elif is_upper_bounded:
+        lognorm = upper_logcdf
+
+    logcdf_numerator = logdiffexp(logcdf, lower_logcdf) if is_lower_bounded else logcdf
+    logcdf_trunc = pt.switch(
+        pt.isinf(logcdf_numerator) & pt.isinf(lognorm), 0.0, logcdf_numerator - lognorm
+    )
+
+    if is_lower_bounded:
+        logcdf_trunc = pt.switch(value < lower, -np.inf, logcdf_trunc)
```

Comment:
This line is causing the test to fail, 
```
AssertionError: 
E           Arrays are not almost equal to 6 decimals
E           {'mu': array(-2.1), 'sigma': array(0.01), 'lower': array(-1.), 'upper': array(0.01), 'value': array(-1.)}
E           x and y -inf location mismatch:
E            x: array(0.)
E            y: array(-inf)
```
as `value` == `lower`. If I change to `<=` then the tests pass. Is there a reason why it is `<` and not `<=`?

Reply:
Perhaps for discrete truncated variables?

Reply:
What do you mean this line is causing it to fail? Sounds like you mean that that line could make it work, that's different. If you commented it out it would still fail no?

Reply:
Is the problem in the `logcdf_numerator` / `lognorm` already? I see you did something with those

Reply:
Applogies I haven't been very clear, there seems to be two problems. With this test,
```python
check_logcdf(
        pm.TruncatedNormal,
        R,
        {"mu": R, "sigma": Rplus, 'lower': Rminusbig, 'upper':Rplus},
        lambda value, mu, sigma, lower, upper: scipy.stats.truncnorm((lower - mu) / sigma, (upper - mu) / sigma, mu, sigma,).logcdf(value),
        decimal=select_by_precision(float64=5, float32=1),
 )
```
which had the error,
```
AssertionError:  
Arrays are not almost equal to 5 decimals
 {'mu': array(-2.1), 'sigma': array(0.01), 'lower': array(-1.), 'upper': array(1.5), 'value': array(0.01)} x and y nan location mismatch:
x: array(nan) 
y: array(-0.)
```
This was caused because `logcdf_numerator` and `lognorm` were both `-inf` on the line,
```python
logcdf_trunc = logcdf_numerator - lognorm
```
So I changed to,
```python
  logcdf_trunc = pt.switch(
        pt.isinf(logcdf_numerator) & pt.isinf(lognorm), 0.0, logcdf_numerator - lognorm
    )
```

What I then found was that this test was also failing,
```python
check_logcdf(
            pm.TruncatedNormal,
            R,
            {"mu": R, "sigma": Rplus, "lower": Rminusbig, "upper": Rplus},
            ft.partial(scipy_logcdf, lower=-np.inf),
            decimal=select_by_precision(float64=6, float32=1),
            skip_paramdomain_outside_edge_test=True,
        )
```
which had the error
```
AssertionError: 
E           Arrays are not almost equal to 6 decimals
E           {'mu': array(-2.1), 'sigma': array(0.01), 'lower': array(-1.), 'upper': array(0.01), 'value': array(-1.)}
E           x and y -inf location mismatch:
E            x: array(0.)
E            y: array(-inf)
```
this led to my thinking in the comment above.

Reply:
scipy does some fancy tricks. 

Since we have a specialized TruncatedNormal and this one is probably the most used one, it may make sense to implement a specialized logcdf and leave the Truncated one alone. For instance I see we have this function, which may be just what we need to get some extra precision? https://github.com/pymc-devs/pymc/blob/02cbac6592506a547117b2daa3f05847d16fe063/pymc/distributions/dist_math.py#L137